### PR TITLE
Handle 'renamed' file notify event

### LINKED ISSUE #2973: LSP mode ignores 'renamed' file watcher event
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [ ] You may also try reproduce the issue using clean environment using the following command
```bash
emacs -q -l lsp-start-plain.el
```
where `lsp-start-plain.el` can be downloaded [here](https://github.com/emacs-lsp/lsp-mode/blob/master/scripts/lsp-start-plain.el).

### Bug description

lsp-mode only handles 'created', 'deleted' and 'changed' file watch events. At least on Windows 10, file watchers as created by file-notify-add-watch raise a 'renamed' event when files are renamed (e.g. through dired). Since this event is currently ignored by lsp-mode, the lsp server is not notified that the file was renamed, and this typically results in the server believing that both the original file and new file exist concurrently, which in itself lead to spurious error reports, double or ambiguous declarations, etc, depending on the language itself.

See https://github.com/emacs-lsp/lsp-mode/blob/fc30baf3d66b9b5c695e851f7e54af926a453c12/lsp-mode.el#L1681 and the documentation from lisp function 'file-notify-add-watch' for references.

### Steps to reproduce

1. Open a file in a lsp workspace (ex: A.cs),
2. C-x C-j (dired-jump),
3. R (dired-rename) and choose a new filename (ex: B.cs),
4. The lsp-server (omnisharp-roslyn in my case) now thinks that both A.cs and B.cs exists.

### Expected behavior

The lsp-server should be notified that A.cs does not exist anymore, and that B.cs is a new file to be managed by the workspace. In particular, the lsp server should not be left thinking that both A.cs and B.cs currently exist.

### Which Language Server did you use?

omnisharp-roslyn

### OS

Windows

### Error callstack

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by thomashilke**
lsp-mode only handles 'created', 'deleted' and 'changed' file watch events. At least on Windows 10, file watchers as created by file-notify-add-watch raise a 'renamed' event when files are renamed (e.g. through dired). Since this event is currently ignored by lsp-mode, the lsp server is not notified that the file was renamed, and this typically results in the server believing that both the original file and new file exist concurrently, which in itself lead to spurious error reports, double or ambiguous declarations, etc, depending on the language itself. 

This commit handles properly the 'rename' event by by notifying the lsp server of the deletion of the file and creation of the file with the new name.

Fixes #2973

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -1679,7 +1679,12 @@ This set of allowed chars is enough for hexifying local file paths.")
      ((and (not (file-directory-p file-name))
            (not (lsp--string-match-any ignored-files file-name))
            (memq event-type '(created deleted changed)))
-      (funcall callback event)))))
+      (funcall callback event))
+     ((and (not (file-directory-p file-name))
+           (not (lsp--string-match-any ignored-files file-name))
+           (memq event-type '(renamed)))
```
Comment (yyoncho) [inline lsp-mode.el:1685]: can you put both memq checks on the first place in the `and`-s?

**Review by yyoncho — COMMENTED** [review]
Looks good. One minor nit:

File: lsp-mode.el (hunk 1 above)
Comment (thomashilke) [reply lsp-mode.el]: Yes sure. Like so?

**Review by yyoncho — APPROVED** [review]

**Review by jcs090218 — APPROVED** [review]

**Comment by thomashilke** [comment]
Some checks fail with the error `Failed to download ‘melpa’ archive.` Is there something that I should do?

**Comment by yyoncho** [comment]
> Some checks fail with the error `Failed to download ‘melpa’ archive.` Is there something that I should do?

No.

Thank you for contributing to `lsp-mode`!

**Pull request merged**